components/script/dom/gpubuffer.rs

```diff
@@ -255,11 +244,65 @@ impl GPUBufferMethods for GPUBuffer {
         }
 
         self.state.set(GPUBufferState::MappingPending);
-        self.map_mode.set(Some(mode));
-        *self.mapping_range.borrow_mut() = map_range;
+        *self.map_info.borrow_mut() = Some(GPUBufferMapInfo {
+            mapping: Rc::new(RefCell::new(Vec::with_capacity(0))),
+            mapping_range: map_range,
+            mapped_ranges: Vec::new(),
+            js_buffers: Vec::new(),
+            map_mode: Some(mode),
+        });
+        *self.map_promise.borrow_mut() = Some(promise.clone());
         promise
     }
 
+    /// https://gpuweb.github.io/gpuweb/#dom-gpubuffer-getmappedrange
+    #[allow(unsafe_code)]
+    fn GetMappedRange(
+        &self,
+        cx: JSContext,
+        offset: GPUSize64,
+        size: GPUSize64,
+    ) -> Fallible<NonNull<JSObject>> {
+        let m_end = if size == 0 { self.size } else { offset + size };
+        let mut info = self.map_info.borrow_mut();
+        let m_info = info.as_mut().unwrap();
+
+        let mut valid = match self.state.get() {
+            GPUBufferState::Mapped | GPUBufferState::MappedAtCreation => true,
+            _ => false,
+        };
+        valid &= offset % RANGE_OFFSET_ALIGN_MASK == 0 &&
+            (m_end - offset) % RANGE_SIZE_ALIGN_MASK == 0 &&
+            offset >= m_info.mapping_range.start &&
+            m_end <= m_info.mapping_range.end;
+        valid &= m_info
+            .mapped_ranges
+            .iter()
+            .all(|range| range.start >= m_end || range.end <= offset);
+        if !valid {
+            return Err(Error::Operation);
+        }
+
+        unsafe extern "C" fn free_func(_contents: *mut c_void, free_user_data: *mut c_void) {
+            let _ = Rc::from_raw(free_user_data as _);
+        }
+
+        let array_buffer = unsafe {
+            NewExternalArrayBuffer(
+                *cx,
+                (m_end - offset) as usize,
+                m_info.mapping.borrow_mut()[offset as usize..m_end as usize].as_mut_ptr() as _,
+                Some(free_func),
+                Rc::into_raw(m_info.mapping.clone()) as _,
```

Comment:
I was looking at this piece a bit. So this would keep the memory for the ArrayBuffer around until it's GC-ed.
I think it's totally fine to land in this stage. Just wanted to note that for the end game, we probably want to "detach" the array buffer on `unmap()`, so that any access to its contents would be invalid and trigger an exception. Perhaps, we could leave a comment somewhere about this TODO

Reply:
We already detach the actual array buffers(which are the only ones exposed to the user) on unmap. We also drop the `Rc` pointer for this array buffer in during dispose `free_func` and set the `map_info` to `None` (so we drop all `Rc` pointers on unmap). What else are we missing here? Should the `contents` be set to `ptr::null()` in `free_func`?